File: src/moai_adk/resources/templates/.claude/hooks/moai/pre_write_guard.py
#!/usr/bin/env python3
# @SEC:PRE-WRITE-GUARD-011
"""
Optimized PreToolUse guard for MoAI-ADK - v0.2.0
Minimal security checking with essential protection.
"""

import json
import sys

# Essential security patterns
SENSITIVE_KEYWORDS = (".env", "/secrets", "/.git/", "/.ssh")
PROTECTED_PATHS = (".moai/memory/",)

def check_file_safety(file_path: str) -> bool:
    """Check if file is safe to edit"""
    if not file_path:
        return True

    path_lower = file_path.lower()

    # Block sensitive patterns
    for keyword in SENSITIVE_KEYWORDS:
        if keyword in path_lower:
            return False

    # Block protected paths
    for protected in PROTECTED_PATHS:
        if protected in file_path:
            return False

    return True

def main() -> None:
    """Main entry point for Claude Code hook system"""
    try:
        data = json.load(sys.stdin)
        tool_name = data.get("tool_name", "")

        if tool_name not in {"Write", "Edit", "MultiEdit"}:
            sys.exit(0)

        tool_input = data.get("tool_input", {}) or {}
        file_path = (tool_input.get("file_path") or
                    tool_input.get("filePath") or
                    tool_input.get("path"))

        if not check_file_safety(str(file_path) if file_path else ""):
            print("BLOCKED: 민감한 파일은 편집할 수 없습니다.", file=sys.stderr)
            sys.exit(2)

        sys.exit(0)

    except Exception:
        # Silent failure to avoid breaking Claude Code session
        sys.exit(0)

if __name__ == "__main__":
    main()

File: moai-adk-ts/src/claude/hooks/pre-write-guard.ts
/**
 * @CODE:HOOK-002 | @CODE:HOOKS-REFACTOR-001 |
 * Related: @CODE:HOOK-002:API, @CODE:PREWRITE-001
 *
 * Pre-Write Guard Hook
 * 파일 쓰기 전 위험 패턴 검증 및 차단
 */

import type { HookInput, HookResult, MoAIHook } from '../types';
import { PROTECTED_PATHS, SENSITIVE_KEYWORDS } from './constants';
import { runHook } from './base';
import { extractFilePath } from './utils';

// Re-export types for test compatibility
export type { HookInput, HookResult } from '../types';

/**
 * Pre-Write Guard Hook - TypeScript port of pre_write_guard.py
 */
export class PreWriteGuard implements MoAIHook {
  name = 'pre-write-guard';

  async execute(input: HookInput): Promise<HookResult> {
    const toolName = input.tool_name;

    // Only check write operations
    if (!toolName || !['Write', 'Edit', 'MultiEdit'].includes(toolName)) {
      return { success: true };
    }

    const toolInput = input.tool_input || {};
    const filePath = extractFilePath(toolInput);

    if (!this.checkFileSafety(filePath || '')) {
      return {
        success: false,
        blocked: true,
        message: '민감한 파일은 편집할 수 없습니다.',
        exitCode: 2,
      };
    }

    return { success: true };
  }

  /**
   * Check if file is safe to edit
   */
  private checkFileSafety(filePath: string): boolean {
    if (!filePath) {
      return true;
    }

    const pathLower = filePath.toLowerCase();

    // Check sensitive keywords
    for (const keyword of SENSITIVE_KEYWORDS) {
      if (pathLower.includes(keyword)) {
        return false;
      }
    }

    // Check protected paths (except templates)
    const isTemplate = filePath.includes('/templates/.moai/');
    if (!isTemplate) {
      for (const protectedPath of PROTECTED_PATHS) {
        if (filePath.includes(protectedPath)) {
          return false;
        }
      }
    }

    return true;
  }
}

// Execute if run directly
if (require.main === module) {
  runHook(PreWriteGuard).catch(() => {
    // Silent failure to avoid breaking Claude Code session
    process.exit(0);
  });
}
